feat(api-graphql): add WebSocket connection health monitoring

anivar opened the pull request:
#### Description of changes

Add read-only WebSocket connection health diagnostics to `AWSWebSocketProvider`:

- **`getConnectionHealth()`** — returns `WebSocketHealthState` with:
  - `isHealthy`: `true` when connection state is `Connected` AND last keep-alive was received within 65 seconds
  - `connectionState`: current `ConnectionState` enum value
  - `lastKeepAliveTime`: timestamp of most recent keep-alive
  - `timeSinceLastKeepAlive`: milliseconds since last keep-alive

- **`isConnected()`** — checks raw WebSocket `readyState === OPEN`

These methods let consumers detect stale connections and build health indicators without modifying any internal state.

**Removed from original implementation:**
- Broken `reconnect()` method (called `close()` which destroys subscription state and reconnection monitor, making recovery impossible)
- Over-engineered persistent storage (module-level `require()` side effect for AsyncStorage, localStorage I/O on every keep-alive, useless after app restart)
- Unused `WebSocketControl` interface (never implemented)
- `getPersistentConnectionHealth()` (depended on removed persistence)

#### Issue #, if available

N/A

#### Description of how you validated changes

- 10 unit tests covering all connection states, keep-alive staleness, and WebSocket readyState values
- Verified 42 existing `AWSAppSyncRealTimeProvider` tests still pass
- TypeScript compilation clean (no new errors)

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Unit Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

bobbor commented on packages/api-graphql/src/Providers/AWSWebSocketProvider/index.ts:
```diff
@@ -49,6 +49,27 @@ import {
 import { awsRealTimeHeaderBasedAuth } from './authHeaders';
 
+// Platform-safe AsyncStorage import
+let AsyncStorage: any;
```
we have the [`KeyValueStorageInterface`](https://github.com/aws-amplify/amplify-js/blob/db51f91b652c5aa20a0c657cf9352d6987905efc/packages/core/src/types/storage.ts#L4).

It can be (partially) used as a type here. and would be better than `any`

bobbor commented on packages/api-graphql/src/Providers/AWSWebSocketProvider/index.ts:
```diff
@@ -1025,4 +1059,114 @@ export abstract class AWSWebSocketProvider {
+		import('../../types').WebSocketHealthState
+	> {
+		let persistentKeepAliveTime: number | undefined;
+		let _timeSinceLastPersistentKeepAlive: number | undefined;
```
what is the use of the `_`.

there is no need to mark variables this way and it can just be `timeSinceLastPersistentKeepAlive`

bobbor commented on packages/api-graphql/src/Providers/AWSWebSocketProvider/index.ts:
```diff
@@ -1025,4 +1059,114 @@ export abstract class AWSWebSocketProvider {
+			this.connectionState === ConnectionState.Connected &&
+			this.keepAliveTimestamp &&
+			timeSinceLastKeepAlive !== undefined &&
+			timeSinceLastKeepAlive < 65000; // 65 second threshold
```
this can be shortened:

```ts
const isHealthy =
			this.connectionState === ConnectionState.Connected &&
			timeSinceLastKeepAlive < 65000; // 65 second threshold
```

bobbor commented on packages/api-graphql/src/Providers/AWSWebSocketProvider/index.ts:
```diff
@@ -1025,4 +1059,114 @@ export abstract class AWSWebSocketProvider {
+		// Use the more recent timestamp (in-memory vs persistent)
+		const lastKeepAliveTime =
+			Math.max(this.keepAliveTimestamp || 0, persistentKeepAliveTime || 0) ||
+			undefined;
```
why the undefined?
why not have `lastKeepAliveTime` be `0`?

this simplifies `timeSinceLastKeepAlive` and `isHealthy`, too

bobbor reviewed (changes requested):
Thanks for opening the PR.
Please take a look at the comments

anivar commented:
## Updates - Code Quality Improvements & Comprehensive Test Coverage

Hi @bobbor! I've addressed all your previous review feedback and made significant improvements to code quality:

### ✅ All Review Feedback Addressed

1. **Promise handling** - Now using proper `.catch()` for AsyncStorage operations
2. **Type safety** - Using `KeyValueStorageInterface` instead of `any`
3. **Import style** - Proper import statements at the top
4. **Code simplification** - Removed unnecessary complexity
5. **Date handling** - Using `Date.now()` instead of `JSON.stringify(new Date())`

### 🎯 Refactoring Improvements

- **Extracted helper methods** for better maintainability:
  - `calculateHealthState()` - Single source of truth for health calculations
  - `getPersistedKeepAliveTimestamp()` - Reusable storage retrieval
  - `persistKeepAliveTimestamp()` - Centralized storage operations
  - `initializePlatformStorage()` - Testable platform detection

- **Enhanced error handling**:
  - Added concurrency guards to `reconnect()` method
  - Proper try/catch/finally blocks throughout
  - Graceful degradation when storage unavailable

- **Type consistency**:
  - Fixed `WebSocketHealthState` to use non-optional fields
  - Sensible defaults (0 for timestamps, Infinity for never-received)

### 🧪 Comprehensive Test Coverage

Added **21 unit tests** covering all new functionality:
- `getConnectionHealth()` - 3 tests
- `getPersistentConnectionHealth()` - 4 tests  
- `isConnected()` - 5 tests
- `reconnect()` - 6 tests (including concurrency protection)
- Error handling - 3 tests

**All 170 tests passing** ✅

### 📊 Impact

- **No breaking changes** - Fully backward compatible
- **Better maintainability** - DRY principles, extracted helpers
- **Production ready** - Comprehensive test coverage, proper error handling

Ready for re-review! Let me know if there's anything else you'd like me to address.

anivar committed:
feat(api-graphql): add WebSocket connection health monitoring

Add getConnectionHealth() and isConnected() to AWSWebSocketProvider for
real-time connection diagnostics. getConnectionHealth() reports whether the
connection is healthy based on connection state and keep-alive staleness
(65s threshold). isConnected() checks raw WebSocket readyState.

Removed broken reconnect(), over-engineered persistent storage, and unused
WebSocketControl interface from original implementation.

anivar committed:
chore: downgrade changeset to patch
